File: leecode/leetcode-cn/911.在线选举.java
class TopVotedCandidate {

    private int[] times;
    private int[] winners;
    private Map<Integer, Integer> voteCounts; // person -> vote

    public TopVotedCandidate(int[] persons, int[] times) {
        int n = persons.length;

        this.times = times;
        winners = new int[n];
        voteCounts = new HashMap<>();

        int winner = -1;

        for (int i = 0; i < n; i++) {
            int p = persons[i];
            voteCounts.put(p, voteCounts.getOrDefault(p, 0) + 1);
            if (voteCounts.get(p) >= voteCounts.getOrDefault(winner, 0)) {
                winner = p;
            }
            winners[i] = winner;
        }
    }

    public int q(int t) {
        int left = 0, right = times.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            if (times[mid] == t)
                return winners[mid];
            else if (times[mid] > t)
                right = mid - 1;
            else
                left = mid + 1;
        }

        return left > 0 ? winners[left - 1] : -1;
    }
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj.q(t);
 */

File: leecode/leetcode-cn/911.在线选举.cpp
/*
 * @lc app=leetcode.cn id=911 lang=cpp
 *
 * [911] 在线选举
 */

// @lc code=start
class TopVotedCandidate {
    vector<pair<int, int>> tops;  // timepoint -> person (who gets the most votes)
public:
    TopVotedCandidate(vector<int>& persons, vector<int>& times) {
        int n = persons.size();
        tops.resize(n);
        pair<int, int> top;  // vote -> person
        unordered_map<int, int> votes;
        for (int i = 0; i < n; ++i) {
            int p = persons[i], t = times[i];
            int vote = ++votes[p];
            if (vote >= top.first)
                top = make_pair(vote, p);
            tops[i] = make_pair(t, top.second);
        }
    }

    int q(int t) {
        // {t, INT_MAX} 不可能达到的上界
        auto i = ranges::upper_bound(tops, make_pair(t, INT_MAX));
        // 由题意，一定可以找到，故不需要判断边界条件
        return (i - 1)->second;
    }
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */
// @lc code=end
